geo: audit functions for duplicate points in Geometry

otan opened the pull request:
Audited all the existing functions to check they work for duplicate
point cases (e.g. `LINESTRING(1.0 1.0, 1.0 1.0, 2.0 2.0)`. The ones I
found that were handled incorrect were distance. In particular, distance
is completely wrong for s2's Distance calculator, so I've opted to
replace that with our own in house instead.

Added a few small tests. If more are incorrect, it will be up to
randomized testing to handle it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49618)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49618#-:-M8RVYBp5FnazS41RzxR:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geogfn/distance.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49618#-M8RTj36DQoBY97hDmus:-M8RTj36DQoBY97hDmut:bgux1th) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7a41036de2306eee0483ea9de984aa40be7fe5/pkg/geo/geogfn/distance.go#L217)):*
> ```Go
> }
> 
> // geographyMinDistanceUpdater finds the minimum distance using spheroid calculations.
> ```

nit: "using spheroid calculations" is stale
___
*[pkg/geo/geomfn/distance.go, line 494 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49618#-M8RVT8NBpx436SnR2xb:-M8RVT8OD2t3g-s5IIMt:bqbz2y9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab7a41036de2306eee0483ea9de984aa40be7fe5/pkg/geo/geomfn/distance.go#L494)):*
> ```Go
> 	p := pointInterface.(*geomGeodistPoint).Coord
> 
> 	// Edge is the same point. Closest point must be any edge vertex.
> ```

nit: Edge is a single point. Closest ...


<!-- Sent from Reviewable.io -->

otan committed:
geo: audit functions for duplicate points in Geometry

Audited all the existing functions to check they work for duplicate
point cases (e.g. `LINESTRING(1.0 1.0, 1.0 1.0, 2.0 2.0)`. The ones I
found that were handled incorrect were distance. In particular, distance
is completely wrong for s2's Distance calculator, so I've opted to
replace that with our own in house instead.

Added a few small tests. If more are incorrect, it will be up to
randomized testing to handle it.

Release note: None

otan commented:
bors r=sumeerbhola
